sharding/peerTypeProvider.go

```diff
@@ -0,0 +1,69 @@
+package sharding
+
+import (
+	"bytes"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/core/check"
+)
+
+// PeerTypeProvider handles the computation of a peer type
+type PeerTypeProvider struct {
+	nodesCoordinator NodesCoordinator
+	epochHandler     EpochHandler
+}
+
+// NewPeerTypeProvider will return a new instance of PeerTypeProvider
+func NewPeerTypeProvider(
+	nodesCoordinator NodesCoordinator,
+	epochHandler EpochHandler,
+) (*PeerTypeProvider, error) {
+	if check.IfNil(nodesCoordinator) {
+		return nil, ErrNilNodesCoordinator
+	}
+	if check.IfNil(epochHandler) {
+		return nil, ErrNilEpochHandler
+	}
+
+	return &PeerTypeProvider{
+		nodesCoordinator: nodesCoordinator,
+		epochHandler:     epochHandler,
+	}, nil
+}
+
+// ComputeForPubKey returns the peer type for a given public key and shard id
+func (ptp *PeerTypeProvider) ComputeForPubKey(pubKey []byte, shardID uint32) (core.ValidatorList, error) {
```

Comment:
this seems too compute intensive to execute for each node every 30 seconds, so I suggest you add a cache for the type with the public key as key, and populate this cache on a new epoch trigger.

all requests that are not found in the cache return "observer" and there's no need to add observers to the cache as it may allow for out of memory vectors.

Reply:
added a cache